src/libraries/System.Linq/src/System/Linq/DefaultIfEmpty.SpeedOpt.cs

```diff
@@ -18,10 +18,21 @@ public TSource[] ToArray()
 
             public List<TSource> ToList()
             {
-                List<TSource> list = _source.ToList();
-                if (list.Count == 0)
+                List<TSource> list;
+                if (_source is IIListProvider<TSource> listProvider && listProvider.GetCount(onlyIfCheap: true) == 0)
                 {
-                    list.Add(_default);
+                    // When _source is empty, the ToList() code path will generally allocate an empty list which must then
+                    // be resized to accept the default item. When it is cheap to determine that it will be empty, directly
+                    // allocate a single item list instead.
+                    list = ToSingleItemList(_default);
```

Comment:
This could also end up being a regression.  Imagine the caller did:
```C#
List<T> list = source.DefaultIfEmpty().ToList();
list.Add(Something());
```
Now that Add could force it to resize whereas previously there may have been room remaining in the initial allocation it created.  I'm not certain this is the right tradeoff, in which case I'd err on the side of "just keep the code simple" and not change this at all.

Reply:
This makes sense, reverted